File: geekbang/algorithm/source-code/1/lc155 最小栈.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int val) {
        s.push(val);
        if (preMin.empty()) preMin.push(val);
        else preMin.push(min(preMin.top(), val));
    }
    
    void pop() {
        s.pop();
        preMin.pop();
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return preMin.top();
    }

private:
    stack<int> s;
    stack<int> preMin;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: geekbang/algorithm/source-code/1/lc155 最小栈.java
class MinStack {

    public MinStack() {
        s = new Stack<>();
        preMin = new Stack<>();
    }
    
    public void push(int val) {
        s.push(val);
        if (preMin.empty()) preMin.push(val);
        else preMin.push(Math.min(preMin.peek(), val));
    }
    
    public void pop() {
        s.pop();
        preMin.pop();
    }
    
    public int top() {
        return s.peek();
    }
    
    public int getMin() {
        return preMin.peek();
    }

    Stack<Integer> s;
    Stack<Integer> preMin;
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */